node/src/api/middleware/authentication.rs

```diff
@@ -0,0 +1,117 @@
+use crate::api::{errors::ApiError, models::AccessToken};
+use actix_http::error::ResponseError;
+use actix_service::Service;
+use actix_web::{
+    dev::{MessageBody, Payload, ServiceRequest, ServiceResponse, Transform},
+    error,
+    FromRequest,
+    HttpRequest,
+};
+use futures::future::{ok, Ready};
+use std::{
+    borrow::BorrowMut,
+    cell::RefCell,
+    future::Future,
+    pin::Pin,
+    rc::Rc,
+    task::{Context, Poll},
+};
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct AuthenticationContext {
+    pubkey: String,
+}
+
+pub trait RequestAuthenticationContext {
+    fn authentication_context(&self) -> Result<AuthenticationContext, ApiError>;
+}
+
+impl RequestAuthenticationContext for HttpRequest {
+    fn authentication_context(&self) -> Result<AuthenticationContext, ApiError> {
+        let access_token = AccessToken::from_request(&self, &mut Payload::None).into_inner()?;
+        Ok(AuthenticationContext {
+            pubkey: access_token.sub,
+        })
+    }
+}
+
+pub struct Authentication;
+
+impl Authentication {
+    pub fn new() -> Authentication {
+        Authentication {}
+    }
+}
+
+impl<S, B> Transform<S> for Authentication
+where
+    S: Service<Request = ServiceRequest, Response = ServiceResponse<B>, Error = error::Error> + 'static,
+    B: MessageBody,
+{
+    type Error = S::Error;
+    type Future = Ready<Result<Self::Transform, Self::InitError>>;
+    type InitError = ();
+    type Request = S::Request;
+    type Response = S::Response;
+    type Transform = AuthenticationService<S>;
+
+    fn new_transform(&self, service: S) -> Self::Future {
+        ok(AuthenticationService::new(service))
+    }
+}
+
+#[derive(Clone)]
+pub struct AuthenticationService<S> {
+    service: Rc<RefCell<S>>,
+}
+
+impl<S> AuthenticationService<S> {
+    fn new(service: S) -> Self {
+        Self {
+            service: Rc::new(RefCell::new(service)),
+        }
+    }
+}
+
+impl<S, B> Service for AuthenticationService<S>
+where
+    S: Service<Request = ServiceRequest, Response = ServiceResponse<B>, Error = error::Error> + 'static,
+    B: MessageBody,
+{
+    type Error = S::Error;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>>>>;
+    type Request = S::Request;
+    type Response = S::Response;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.service.poll_ready(cx).map_err(error::Error::from)
+    }
+
+    fn call(&mut self, request: Self::Request) -> Self::Future {
+        let mut service = self.service.clone();
+
+        // Ignore requests to the status endpoint
+        if request.uri() == "/status" {
+            Box::pin(async move { service.borrow_mut().call(request).await })
```

Comment:
there is a trick here, you need to make sure that RefMut is dropped outside of the Future context, otherwise theremight be a panic (e.g. from another call to RefCell borrow_mut()):
```suggestion
            let fut = service.borrow_mut().call(request);
            Box::pin(async move { fut.await })
```